### PR TITLE
[CDAP-17748] One pager for hierarchy widget in UI

### PR DESCRIPTION
**Pull request by vjosadika**
JIRA: https://cdap.atlassian.net/browse/CDAP-17748
https://cdap.atlassian.net/browse/CDAP-17749

### CODE REVIEW

File: cdap-ui/app/cdap/components/AbstractWidget/HierarchyWidget/__tests__/HierarchyWidget.tests.ts [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an 'AS IS' BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+import { typeHandler, flatToTree, addPathToFlatMap, toJsonHandler } from '../dataHandler';
+
+describe('Unit tests for Hierarchy Widget', () => {
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/AbstractWidget/HierarchyWidget/__tests__/HierarchyWidget.tests.ts]: nit: This is unit test for data handler helper function in Hierarchy widget right? We are not necessarily testing the hierarchy widget.

File: cdap-ui/app/cdap/components/AbstractWidget/HierarchyWidget/__tests__/HierarchyWidget.tests.ts [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+      name: 'Id',
+      type: 'boolean',
+    };
+    const output = {
+      name: 'Id',
+      type: 'boolean',
+      children: [],
+    };
+    const result = typeHandler(field);
+    expect(JSON.stringify(result)).toBe(JSON.stringify(output));
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/AbstractWidget/HierarchyWidget/__tests__/HierarchyWidget.tests.ts]: Do we need to stringify the JSON? Can we use `expect(result).to.deep.equal(output)`? Can we use `.toEqual` provided by jest?

File: cdap-ui/app/cdap/components/AbstractWidget/HierarchyWidget/__tests__/HierarchyWidget.tests.ts [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+  });
+
+  it('Should handle array type', () => {
+    const field = {
+      name: 'Id',
+      type: {
+        type: 'array',
+        items: 'string',
+      },
+      children: ['string'],
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/AbstractWidget/HierarchyWidget/__tests__/HierarchyWidget.tests.ts]: The input field will not have a children key right? The avro schema spec defines only `name` and `type`?

File: cdap-ui/app/cdap/components/AbstractWidget/HierarchyWidget/__tests__/HierarchyWidget.tests.ts [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+      children: [{ name: 'Name', type: 'string' }],
+    };
+    const result = typeHandler(field);
+    expect(JSON.stringify(result)).toBe(JSON.stringify(output));
+  });
+
+  it('Should handle union type', () => {
+    const field = {
+      name: 'Id',
+      type: ['string'],
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/AbstractWidget/HierarchyWidget/__tests__/HierarchyWidget.tests.ts]: Can the test include a union of multiple types like `int, float and double`?

File: cdap-ui/app/cdap/components/AbstractWidget/HierarchyWidget/__tests__/HierarchyWidget.tests.ts [hunk 5]
```diff
@@ -0,0 +1,419 @@
...
+    const field = {
+      name: 'Id',
+      type: {
+        type: 'bytes',
+        logicalType: 'decimal',
+        precision: 32,
+        scale: 3,
+      },
+    };
+    const output = {
```
Comment (ajainarayanan) [inline cdap-ui/app/cdap/components/AbstractWidget/HierarchyWidget/__tests__/HierarchyWidget.tests.ts:254]: Does this mean we are not capturing the precision and scale attributes?

**Review by ajainarayanan — COMMENTED** [review]
Thanks for adding tests @vjosadika .

Can we add a few more complex cases for converting a flat schema to a tree and vice versa?  This is to test some of the complex types we tested during the development. 

For eg: If a top level record has 10 fields with random mix of  complex types, do we generate the right JSON expected by the backend.
And if given the JSON from backend do we generate the the flat fields back in the right way.

**Comment by aviachar** [comment]
Closing old PRs

**Pull request closed without merging**